presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -14,25 +14,22 @@
 package com.facebook.presto.sql.planner.iterative;
 
 import com.facebook.presto.Session;
-import com.facebook.presto.matching.Matchable;
+import com.facebook.presto.matching.Captures;
 import com.facebook.presto.matching.Pattern;
 import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
 import com.facebook.presto.sql.planner.SymbolAllocator;
 import com.facebook.presto.sql.planner.plan.PlanNode;
 
 import java.util.Optional;
 
-public interface Rule extends Matchable
+public interface Rule<T>
```

Comment:
This is so that the node passed to `Rule#apply` has a concrete type, determined by the Rule's `Pattern<T>`.

Not having `<T>` in Rule means the body of the apply method needs to cast the passed PlanNode to a concrete subtype. This has the drawbacks of:
1. being unsafe,
2. re-asserting the type that's been expressed in the Pattern already,
3. being cumbersome.

I think the biggest challenge when working with `Rule<T>` is getting the 'same T' for both `Rule` and `Pattern`, while the `Rule` is only known as `Rule<?>` (e.g. b/c it's stored in a `Collection<Rule<?>>`). It struck me today that all that's needed is passing the `Rule<?>` to a method parameterized with the T type and doing the work (including getting the Pattern<T>) within the method. 

I'd also say the challenges will only be in a couple of places in framework/planner code, while the user/rule code actually gets simpler.

See `RuleAssert#applyRule` for an example of the approach I metnioned above, and `IterativeOptimier#transform` for another - and probably the last one ever :))